rust/src/derivation.rs

```diff
@@ -0,0 +1,539 @@
+        network: zcash_address::Network,
+        _data: [u8; 20],
+    ) -> Result<Self, ConversionError<Self::Error>> {
+        Ok(AddressMetadata {
+            network,
+            addr_type: AddressType::P2sh,
+        })
+    }
+
+    fn try_from_tex(
+        network: zcash_address::Network,
+        _data: [u8; 20],
+    ) -> Result<Self, ConversionError<Self::Error>> {
+        Ok(AddressMetadata {
+            network,
+            addr_type: AddressType::Tex,
+        })
+    }
+}
+
+/// Returns the network type and address kind for the given address string,
+/// if the address is a valid Zcash address.
+///
+/// Address kind codes are as follows:
+/// * p2pkh: 0
+/// * p2sh: 1
+/// * sapling: 2
+/// * unified: 3
+/// * tex: 4
+///
+/// # Safety
+///
+/// - `address` must be non-null and must point to a null-terminated UTF-8 string.
+/// - The memory referenced by `address` must not be mutated for the duration of the function call.
+#[no_mangle]
+pub unsafe extern "C" fn zcashlc_get_address_metadata(
+    address: *const c_char,
+    network_id_ret: *mut u32,
+    addr_kind_ret: *mut u32,
+) -> bool {
+    let res = catch_panic(|| {
+        let addr = unsafe { CStr::from_ptr(address).to_str()? };
+        let zaddr = ZcashAddress::try_from_encoded(addr)?;
+
+        // The following .unwrap is safe because address type detection
+        // cannot fail for valid ZcashAddress values.
+        let addr_meta: AddressMetadata = zaddr.convert().unwrap();
+        unsafe {
+            *network_id_ret = match addr_meta.network {
+                zcash_address::Network::Main => 1,
+                zcash_address::Network::Test => 0,
+                zcash_address::Network::Regtest => {
+                    return Err(anyhow!("Regtest addresses are not supported."));
+                }
+            };
+
+            *addr_kind_ret = match addr_meta.addr_type {
+                AddressType::P2pkh => 0,
+                AddressType::P2sh => 1,
+                AddressType::Sapling => 2,
+                AddressType::Unified => 3,
+                AddressType::Tex => 4,
+                AddressType::Sprout => {
+                    return Err(anyhow!("Sprout addresses are not supported."));
+                }
+            };
+        }
+
+        Ok(true)
+    });
+    unwrap_exc_or(res, false)
+}
+
+/// Extracts the typecodes of the receivers within the given Unified Address.
+///
+/// Returns a pointer to a slice of typecodes. `len_ret` is set to the length of the
+/// slice.
+///
+/// See the following sections of ZIP 316 for details on how to interpret typecodes:
+/// - [List of known typecodes](https://zips.z.cash/zip-0316#encoding-of-unified-addresses)
+/// - [Adding new types](https://zips.z.cash/zip-0316#adding-new-types)
+/// - [Metadata Items](https://zips.z.cash/zip-0316#metadata-items)
+///
+/// # Safety
+///
+/// - `ua` must be non-null and must point to a null-terminated UTF-8 string containing an
+///   encoded Unified Address.
```

Comment:
Is that a safety requirement, or a correctness requirement? I think it's a correctness requirement, and a null pointer will be returned if it is not a correct UA.
```suggestion
/// - `ua` must be non-null and must point to a null-terminated UTF-8 string.
/// - The memory referenced by `ua` must not be mutated for the duration of the function call.
```